compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp

```diff
@@ -0,0 +1,148 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/GlobalOptimization/PassDetail.h"
+#include "iree/compiler/GlobalOptimization/Passes.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/SCF/Transforms/Transforms.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/TopologicalSortUtils.h"
+
+#include "llvm/Support/Debug.h"
+
+using namespace mlir;
+
+#define DEBUG_TYPE "global-loop-invariant-code-motion"
+#define LICM_DBGS() (llvm::dbgs() << '[' << DEBUG_TYPE << "] ")
+
+// Check if the op and its producers are loop invariants and hoistable. Results
+// are cached in hoistableOpMap.
+//
+// It expects the op's producers have been checked and the results exist in
+// `hoistableOpMap`.
+static bool
+isHoistableOp(LoopLikeOpInterface loopOp, Operation *op,
+              llvm::SmallDenseMap<Operation *, bool> &hoistableOpMap) {
+  // Currently we don't handle implicit captures, so don't hoist ops with
+  // regions.
+  if (op->getNumRegions() > 0) {
+    hoistableOpMap[op] = false;
+    return false;
+  }
+
+  // Check if the op type is hoistable.
+  if (!isa<tensor::EmptyOp, tensor::PackOp>(op)) {
+    hoistableOpMap[op] = false;
+    return false;
+  }
+
+  // Check if all producers are hoistable.
+  bool hoistable = true;
+  for (OpOperand &operand : op->getOpOperands()) {
+    Value value = operand.get();
+    // Ignore values defined outside the loop.
+    if (loopOp.isDefinedOutsideOfLoop(value))
+      continue;
+
+    Operation *producer = value.getDefiningOp();
+    // If the producer is not an operation, can't hoist it.
+    if (!producer) {
+      hoistable = false;
+      break;
+    }
+
+    auto producerResult = hoistableOpMap.find(producer);
+    assert(producerResult != hoistableOpMap.end() &&
+           "producer should have been checked");
+    if (!producerResult->second) {
+      hoistable = false;
+      break;
+    }
+  }
+
+  hoistableOpMap[op] = hoistable;
+  return hoistable;
+}
+
+static LogicalResult hoistLoopInvariants(LoopLikeOpInterface loopOp,
+                                         RewriterBase &rewriter) {
+  // First find the root ops can be hoisted. The root op needs to satisfy:
+  // 1. It is a root op having benefits to be hoisted (e.g. tensor.pack)
+  // 2. Its backward slice can be hoisted (e.g. they are loop invariant)
+  SmallVector<Operation *> opsToHoist;
+  llvm::SmallDenseMap<Operation *, bool> hoistableOpMap;
+  for (Region *region : loopOp.getLoopRegions()) {
+    // Consider only the top-level ops in the region.
+    // The forward visiting also ensures we are check and add ops in topological
+    // order.
+    for (Operation &op : region->getOps()) {
```

Comment:
I think you actually need to walk the ops per region... Just iterating might not always work. In this case we have only one block, so it works. So either check for a single block and dont do anything if it isnt a single block region, then you can just iterate over ops in the block.
If you want to handle multiple blocks, you need to walk the IR to make sure you are getting producers before consumers.

Reply:
We can add checks but current SCF loops can only have one block per region, I think, and we don't support hoisting from ops that are nested into other region ops. The only loop we gather right now (`scf.while`) has two regions, which are listed in topological order. I don't think we can do much from the `LoopLikeInterfaceAPI` right now. We may want to revisit this as we add more loop types, if that ever happens.

Reply:
> We can add checks but current SCF loops can only have one block per region, I think, and we don't support hoisting from ops that are nested into other region ops. The only loop we gather right now (`scf.while`) has two regions, which are listed in topological order. I don't think we can do much from the `LoopLikeInterfaceAPI` right now. We may want to revisit this as we add more loop types, if that ever happens.

LoopLikeOpInterface doesn't check for a single block does it? This is written on the interface. If you don't explicitly abort on multiple blocks it can cause crashes for ops that implement the interface and have multiple blocks

Reply:
For simplicity now, I added a check to make sure loop's region has only single block.